juju/juju
# Handle local charm resources

The following attempts to clarify what resources should use which
resource client depending on the charm URL.

Previously, before charmhub client was introduced, all charms went
through the use of the local cache (renamed to resourceState) and if it
wasn't there then an attempt to connect to the charmstore client was
made. With the introduction of charmhub this also followed the same 
pattern. Except for local charms, those were never served correctly.
Instead, they were blackholed.

The fix here is to explicitly handle local charms by stating that no
outbound connection is ever used for surfacing a resource. This will
ensure that when deploying a local resource we show the error to the
user in a much more timely manner.

Additionally, in this commit is the re-working/re-naming of all the
indirections to help make more sense. As somebody first coming to
resources and trying to understand the indirections, the naming of these
things were terrible!

"Cache" was used for accessing state, instead, a better word 
"repository" is used to show that it doesn't just serve the data from a 
local cache as it infers by the name.

Naming is hard, but hopefully, this is better for others reading the
code.


## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ./acceptancetests/repository/charms/dummy-resource  --resource foo="./acceptancetests/repository/charms/dummy-resource/foo.txt" --resource bar="./acceptancetests/repository/charms/dummy-resource/bar.txt"
$ watch juju status
```

Ensure that the foo/bar resources are deployed.
